assets/javascripts/search.js

```diff
@@ -0,0 +1,218 @@
+
+function debounce(fn) {
+  let timeout;
+  return function() {
+    const args = Array.prototype.slice.call(arguments);
+    const context = this;
+    clearTimeout(timeout);
+    timeout = setTimeout(function() {
+      fn.apply(context, args);
+    }, 100);
+  };
+}
+
+class LunrSearch {
+  constructor(elem, options) {
+    this.search_elem = elem;
+    if (options.query) {
+      this.search_elem.val(options.query);
+    }
+    this.quickSearchResults = $(options.quickSearchResults);
+    this.quickSearchEntries = $(
+      options.quickSearchEntries,
+      this.quickSearchResults
+    );
+    this.searchResults = $(options.searchResults);
+    this.searchEntries = $(options.searchEntries, this.searchResults);
+    this.quickSearchTemplate = this.compileTemplate(
+      $(options.quickSearchTemplate)
+    );
+    this.searchTemplate = this.compileTemplate($(options.searchTemplate));
+    this.searchMoreButton = $(".search-more-button");
+    this.searchSpinner = $(".search-spinner");
+    this.searchHeader = $(".search-header");
+    this.searchWorker = new Worker("/assets/javascripts/search_worker.js");
+
+    this.reindex(options.version, options.indexUrl);
+
+    this.bindQuicksearchKeypress();
+    this.bindQuicksearchBlur();
+    this.bindQuicksearchFocus();
+
+    this.searchWorker.addEventListener("message", event => {
+      if (event.data.type.indexed) {
+        this.populateSearchFromQuery();
+      } else {
+        if (event.data.type.quicksearch) {
+          this.displayQuicksearch(event.data.query, event.data.results);
+        } else {
+          this.displaySearchResults(event.data.query, event.data.results);
+        }
+      }
+    });
+    this.searchMoreButton.on("click", () => {
+      if (this.page) {
+        this.page = this.page + 1;
+      } else {
+        this.page = 2;
+      }
+    });
+  }
+
+  compileTemplate($template) {
+    const template = $template.text();
+    Mustache.parse(template, "[[ ]]");
+    return function(view, partials) {
+      return Mustache.render(template, view, partials);
+    };
+  }
+
+  reindex(version, indexDataUrl) {
+    this.version = version;
+    $.getJSON(indexDataUrl, data => {
+      data.type = { index: true };
+      this.searchWorker.postMessage(data);
+    });
+  }
+
+  bindQuicksearchKeypress() {
+    this.search_elem.bind(
+      "keyup",
+      debounce(() => {
+        this.search(this.search_elem.val());
+      })
+    );
+  }
+
+  bindQuicksearchFocus() {
+    this.search_elem.bind(
+      "focus",
+      debounce(() => {
+        if (this.search_elem.val()) {
+          this.quickSearchResults.show();
+        }
+      })
+    );
+  }
+
+  bindQuicksearchBlur() {
+    this.search_elem.bind(
+      "blur",
+      debounce(() => {
+        this.quickSearchResults.hide();
+      })
+    );
+  }
+
+  bindQuicksearchMousedown() {
+    $(".autocomplete-results a").each(function(_, el) {
+      $(el).bind("mousedown", function(event) {
+        event.preventDefault();
```

Comment:
Just preventing the default and returning? I'm assuming this is so the anchors don't try to visit their hrefs. Should we use a different element like `<button>` in that case to be more semantically accurate?

Reply:
I think that's a valid change but probably should be put off until after this is deployed. Would you mind opening an issue?

Reply:
Opened #550